File: update_summary.py
#!/usr/bin/env python3
"""
Script to generate SUMMARY.md based on the directory structure under /src
"""
import os
from pathlib import Path
from typing import List, Tuple, Optional

def get_display_name(filename: str) -> str:
    """Convert filename to display name"""
    # Remove .md extension
    name = filename.replace('.md', '')
    # For README files, use the parent directory name
    if name == 'README':
        return None
    return name

def process_directory(base_path: Path, dir_path: Path, level: int = 0) -> List[str]:
    """Recursively process directory and generate SUMMARY entries"""
    lines = []
    indent = '  ' * level
    
    # Get all items in directory
    try:
        items = sorted(dir_path.iterdir())
    except PermissionError:
        return lines
    
    # Separate files and directories
    md_files = []
    subdirs = []
    
    for item in items:
        if item.is_file() and item.suffix == '.md' and item.name != 'SUMMARY.md':
            md_files.append(item)
        elif item.is_dir():
            subdirs.append(item)
    
    # Process subdirectories
    for subdir in subdirs:
        # Check if there's a README.md in the subdirectory
        readme_path = subdir / 'README.md'
        relative_path = readme_path.relative_to(base_path)
        
        # Use directory name as the title
        title = subdir.name
        
        if readme_path.exists():
            lines.append(f'{indent}- [{title}](./{relative_path})')
            # Process files in subdirectory with increased indentation
            subdir_lines = process_directory(base_path, subdir, level + 1)
            lines.extend(subdir_lines)
        else:
            # If no README, still process subdirectory
            lines.append(f'{indent}- [{title}]')
            subdir_lines = process_directory(base_path, subdir, level + 1)
            lines.extend(subdir_lines)
    
    # Process markdown files (excluding README.md as it's already processed)
    for md_file in md_files:
        if md_file.name == 'README.md':
            continue
        
        relative_path = md_file.relative_to(base_path)
        display_name = get_display_name(md_file.name)
        if display_name:
            lines.append(f'{indent}- [{display_name}](./{relative_path})')
    
    return lines

def generate_summary(src_path: Path) -> str:
    """Generate SUMMARY.md content from src directory structure"""
    lines = ['# Summary', '']
    
    # Add aboutMe.md at the top
    about_me = src_path / 'aboutMe.md'
    if about_me.exists():
        lines.append('- [about me](./aboutMe.md)')
        lines.append('')
    
    # Get all subdirectories
    subdirs = [d for d in sorted(src_path.iterdir()) if d.is_dir()]
    
    for subdir in subdirs:
        # Create section header
        section_name = subdir.name.title()
        lines.append(f'# {section_name}')
        lines.append('')
        
        # Process the subdirectory
        subdir_lines = process_directory(src_path, subdir, level=0)
        lines.extend(subdir_lines)
        lines.append('')
    
    return '\n'.join(lines)

def main() -> int:
    # Get the src directory path
    script_dir = Path(__file__).parent
    src_path = script_dir / 'src'
    
    if not src_path.exists():
        print(f"Error: {src_path} does not exist")
        return 1
    
    # Generate SUMMARY content
    summary_content = generate_summary(src_path)
    
    # Write to SUMMARY.md
    summary_path = src_path / 'SUMMARY.md'
    with open(summary_path, 'w', encoding='utf-8') as f:
        f.write(summary_content)
    
    print(f"Successfully updated {summary_path}")
    return 0

if __name__ == '__main__':
    exit(main())

File: update_summary.rs
use std::fs;
use std::io::{self, Write};
use std::path::{Path, PathBuf};

/// Get display name from filename
fn get_display_name(filename: &str) -> Option<String> {
    let name = filename.strip_suffix(".md").unwrap_or(filename);
    if name == "README" {
        None
    } else {
        Some(name.to_string())
    }
}

/// Recursively process directory and generate SUMMARY entries
fn process_directory(base_path: &Path, dir_path: &Path, level: usize) -> io::Result<Vec<String>> {
    let mut lines = Vec::new();
    let indent = "  ".repeat(level);
    
    // Get all items in directory
    let mut entries: Vec<_> = fs::read_dir(dir_path)?
        .filter_map(|e| e.ok())
        .collect();
    
    // Sort entries by name
    entries.sort_by_key(|e| e.file_name());
    
    // Separate files and directories
    let mut md_files = Vec::new();
    let mut subdirs = Vec::new();
    
    for entry in entries {
        let path = entry.path();
        let file_name = entry.file_name();
        let file_name_str = file_name.to_string_lossy();
        
        if path.is_file() && file_name_str.ends_with(".md") && file_name_str != "SUMMARY.md" {
            md_files.push(path);
        } else if path.is_dir() {
            subdirs.push(path);
        }
    }
    
    // Process subdirectories
    for subdir in subdirs {
        let dir_name = subdir.file_name().unwrap().to_string_lossy();
        let readme_path = subdir.join("README.md");
        
        if readme_path.exists() {
            let relative_path = readme_path.strip_prefix(base_path).unwrap();
            let relative_path_str = relative_path.to_str().unwrap().replace('\\', "/");
            lines.push(format!("{}- [{}](./{})", indent, dir_name, relative_path_str));
            
            // Process files in subdirectory with increased indentation
            if let Ok(subdir_lines) = process_directory(base_path, &subdir, level + 1) {
                lines.extend(subdir_lines);
            }
        } else {
            // If no README, still process subdirectory
            lines.push(format!("{}- [{}]", indent, dir_name));
            if let Ok(subdir_lines) = process_directory(base_path, &subdir, level + 1) {
                lines.extend(subdir_lines);
            }
        }
    }
    
    // Process markdown files (excluding README.md as it's already processed)
    for md_file in md_files {
        let file_name = md_file.file_name().unwrap().to_string_lossy();
        if file_name == "README.md" {
            continue;
        }
        
        if let Some(display_name) = get_display_name(&file_name) {
            let relative_path = md_file.strip_prefix(base_path).unwrap();
            let relative_path_str = relative_path.to_str().unwrap().replace('\\', "/");
            lines.push(format!("{}- [{}](./{})", indent, display_name, relative_path_str));
        }
    }
    
    Ok(lines)
}

/// Generate SUMMARY.md content from src directory structure
fn generate_summary(src_path: &Path) -> io::Result<String> {
    let mut lines = vec!["# Summary".to_string(), String::new()];
    
    // Add aboutMe.md at the top
    let about_me = src_path.join("aboutMe.md");
    if about_me.exists() {
        lines.push("- [about me](./aboutMe.md)".to_string());
        lines.push(String::new());
    }
    
    // Get all subdirectories
    let mut subdirs: Vec<_> = fs::read_dir(src_path)?
        .filter_map(|e| e.ok())
        .map(|e| e.path())
        .filter(|p| p.is_dir())
        .collect();
    
    subdirs.sort();
    
    for subdir in subdirs {
        // Create section header
        let dir_name = subdir.file_name().unwrap().to_string_lossy();
        let section_name = capitalize_first(&dir_name);
        lines.push(format!("# {}", section_name));
        lines.push(String::new());
        
        // Process the subdirectory
        if let Ok(subdir_lines) = process_directory(src_path, &subdir, 0) {
            lines.extend(subdir_lines);
        }
        lines.push(String::new());
    }
    
    Ok(lines.join("\n"))
}

/// Capitalize first letter of a string (simple title case)
fn capitalize_first(s: &str) -> String {
    let mut chars = s.chars();
    match chars.next() {
        None => String::new(),
        Some(first) => {
            let mut result = first.to_uppercase().to_string();
            result.push_str(&chars.as_str());
            result
        }
    }
}

fn main() -> io::Result<()> {
    // Get the src directory path
    let src_path = PathBuf::from("src");
    
    if !src_path.exists() {
        eprintln!("Error: {:?} does not exist", src_path);
        std::process::exit(1);
    }
    
    // Generate SUMMARY content
    let summary_content = generate_summary(&src_path)?;
    
    // Write to SUMMARY.md
    let summary_path = src_path.join("SUMMARY.md");
    let mut file = fs::File::create(&summary_path)?;
    file.write_all(summary_content.as_bytes())?;
    
    println!("Successfully updated {:?}", summary_path);
    
    Ok(())
}
